File: go/heap/max_heap.go
package heap

import (
	"../myArray"
	"errors"
)

// 介于堆也要求元素具有可排序性，因此这里使用整型动态数组，而不是泛型数组
type MaxHeap struct {
	data *myArray.MyArray
}

func New(arg ...interface{}) (*MaxHeap, error) {
	if len(arg) == 0 {
		data, err := myArray.NewIntArray(10)
		return &MaxHeap{data: data,}, err
	}

	switch v := arg[0].(type) {
	case int:
		capacity := v
		data, err := myArray.NewIntArray(capacity)
		return &MaxHeap{data: data,}, err
	case []int:
		// 这里要支持 heapify 就必须底层的数组支持传入数组做参数的形式
		length := len(v)
		data, err := myArray.NewIntArray(v)
		heap := &MaxHeap{data: data,}
		lastBranchIndex := (length - 1 - 1) / 2
		for i := lastBranchIndex; i >= 0; i-- {
			heap.SiftDown(i)
		}
		return heap, err
	default:
		return nil, nil
	}
}

func (m *MaxHeap) GetSize() int {
	return m.data.GetLength()
}

func (m *MaxHeap) IsEmpty() bool {
	return m.data.IsEmpty()
}

func (m *MaxHeap) Parent(index int) int {
	// 这里应该不用对 0 进行另外的判断把，应该 -1 / 2 = -0.5，应该是向 0 舍吧，所以还是0
	return (index - 1) / 2
}

func (m *MaxHeap) LeftChild(index int) int {
	return index * 2 + 1
}

func (m *MaxHeap) RightChild(index int) int {
	return index * 2 + 2
}

func (m *MaxHeap) Add(value int) {
	m.data.AddLast(value)
	m.siftUp()
}

func (m *MaxHeap) siftUp() {
	curIndex := m.data.GetLength() - 1
	value, _ := m.data.Get(curIndex)
	parentIndex := m.Parent(curIndex)
	parent, _ := m.data.Get(parentIndex)
	for ; curIndex > 0 && parent < value; {
		m.data.Set(parentIndex, value)
		m.data.Set(curIndex, parent)
		curIndex = parentIndex
		parentIndex = m.Parent(curIndex)
		parent, _ = m.data.Get(parentIndex)
	}
}

func (m *MaxHeap) FindMax() (int, error) {
	return m.data.Get(0)
}

func (m *MaxHeap) RemoveMax() (value int, err error) {
	if m.data.GetLength() == 0 {
		err = errors.New("The Head is Empty")
		return
	}

	value, _ = m.data.Get(0)
	lastIndex := m.data.GetLength() - 1
	lastValue, _ := m.data.Get(lastIndex)
	m.data.Set(0, lastValue)
	m.data.RemoveLast()
	m.SiftDown(0)
	return
}

func (m *MaxHeap) SiftDown(index int) {
	size := m.data.GetLength()
	for ; m.LeftChild(index) < size; {
		li := m.LeftChild(index)
		ri := li + 1
		j := li
		left, _ := m.data.Get(li)
		right, _ := m.data.Get(ri)
		if ri < size && right > left {
			j = ri
		}
		
		max, _ := m.data.Get(j)
		value, _ := m.data.Get(index)
		if value < max {
			m.swap(index, j)
			index = j
		} else {
			break
		}
	}
}

func (m *MaxHeap) swap(x int, y int) {
	v1, _ := m.data.Get(x)
	v2, _ := m.data.Get(y)
	m.data.Set(x, v2)
	m.data.Set(y, v1)
}

func (m *MaxHeap) Replace(value int) error {
	err := m.data.Set(0, value)
	if err != nil {
		return err
	}

	m.SiftDown(0)
	return nil
}

File: js/heap/maxHeap.js
/**
 * 带有容量调整的最大堆
 */
class MaxHeap {
  /**
   * @param {int | Array} param 可以传入一个预估的堆的容量值或者传入一个数组
   */
  constructor(param) {
    if (Array.isArray(param)) {
      this.data = Array(param.length * 2);
      this.size = param.length * 2;
      this.length = param.length;

      for (let i = 0; i < this.length; i++) {
        this.data[i] = param[i];
      }
      this.heapify();
    } else {
      this.data = Array(param);
      this.size = param;
      this.length = 0;
    }
  }

  heapify() {
    const lastForkIndex = this.parent(this.length-1);
    for (let i = lastForkIndex; i >= 0; i--) {
      this.shiftDown(i);
    }
  }

  left(i) {
    return i * 2 + 1;
  }

  right(i) {
    return i * 2 + 2;
  }

  parent(i) {
    return Math.floor((i - 1) / 2);
  }
  
  /**
   * 注意虽然叫 GetSize 但其实是返回了当前堆中元素的个数
   * @returns {int} 堆中元素数量
   */
  GetSize() {
    return this.length;
  }

  IsEmpty() {
    return this.length === 0;
  }

  /**
   * 向堆中添加一个值
   * @param {any} val
   */
  Add(val) {
    if (this.length === this.size) {
      this.resize(this.length * 2);
    }

    this.data[this.length] = val;
    this.shiftUp(this.length);
    this.length++;
  }

  /**
   * 查看堆中最大的元素值
   * @returns {any} 堆中最大的元素
   */
  FindMax() {
    if (this.length === 0) {
      throw new Error("Empty Heap");
    }

    return this.data[0];
  }

  /**
   * 从堆中删除最大的元素并返回
   * @returns {any} 堆中最大的元素
   */
  ExtractMax() {
    const res = this.FindMax();
    this.data[0] = this.data[this.length-1];
    this.data[this.length-1] = null;
    this.length--;

    // 缩容操作
    if (this.length === this.size / 4 && this.size / 2 !== 0) {
      this.resize(this.size / 2);
    }

    this.shiftDown(0);
    return res;
  }
  
  /**
   * 将索引为 index 的元素进行下沉操作
   * @param {int} index
   */
  shiftDown(index) {
    while(this.left(index) < this.size) {
      let maxIndex = this.left(index);
      const leftVal = this.data[maxIndex];
      if (maxIndex + 1 < this.size && this.data[maxIndex+1] > leftVal) {
        maxIndex = maxIndex + 1;
      }

      const maxVal = this.data[maxIndex];
      if (maxVal > this.data[index]) {
        this.swap(maxIndex, index);
        index = maxIndex;
      } else {
        break;
      }
    }
  }

  /**
   * 将索引为 index 的元素进行上浮操作
   * @param {int} index
   */
  shiftUp(index) {
    let parentIndex = this.parent(index);
    while(index > 0 && this.data[parentIndex] < this.data[index]) {
      this.swap(parentIndex, index);
      index = parentIndex;
    }
  }

  /**
   * 交换堆中索引为 i 和 j 的元素的值
   * @param {int} i
   * @param {int} j
   */
  swap(i, j) {
    const temp = this.data[i];
    this.data[i] = this.data[j];
    this.data[j] = temp;
  }

  /**
   * 容量调整
   * @param {int} newSize 新的容量
   */
  resize(newSize) {
    const data = Array(newSize);

    for (let i = 0; i < this.length; i++) {
      data[i] = this.data[i];
    }

    this.size = newSize;
    this.data = data;
  }
}